src/dragonfs.c

```diff
@@ -684,36 +703,93 @@ int dfs_dir_findnext(char *buf)
     return __dfs_findnext(buf, &next_entry);
 }
 
+static uint32_t prime_hash(const char *str, uint32_t prime)
+{
+    uint32_t hash = 0;
+    while(*str) {
+        char c = *str++;
+        hash = (hash * prime) + c;
+    }
+    return hash;
+}
 
-int dfs_open(const char * const path)
+static int32_t lookup_file_hash_entry(uint32_t hash)
 {
-    /* Try to find file */
-    directory_entry_t *dirent;
-    int ret = recurse_path(path, WALK_OPEN, &dirent, TYPE_FILE);
+    int32_t left = 0;
+    int32_t right = file_lookup->num_files-1;
+    int32_t result = -1;
+    while(left <= right) {
+        int32_t mid = left+((right-left)/2);
+        if(file_lookup->files[mid].path_hash == hash) {
+            result = mid;
+            right = mid-1;
+        } else if(file_lookup->files[mid].path_hash < hash) {
+            left = mid+1;
+        } else {
+            right = mid-1;
+        }
+    }
+    return result;
+}
 
-    if(ret != DFS_ESUCCESS)
-    {
-        /* File not found, or other error */
-        return ret;
+static dfs_file_entry_t *lookup_file_path_entry(const char *const path, int32_t start_index, uint32_t hash)
+{
+    size_t buf_len = ROUND_UP(file_lookup->path_buf_size, 16);
+    char alignas(16) path_buf[buf_len];
+    uint32_t src_path_len = strlen(path)+1;
+    if(src_path_len % 2 != 0) {
+        src_path_len++;
+    }
```

Comment:
This shouldn't be required, dma_read handles odd sizes too.

Reply:
Code updated since stored path_len is true path_len.